### DISCUSSION TITLE
Is `nl.adaptivity.xmlutil.serialization.XML` threadsafe?

### DISCUSSION THREAD

**Discussion by felixkrull-neuland** [pdvrieze/xmlutil#245] [Q&A] [answered]
After updating to 0.90.2, I'm getting an intermittent exception when running unit tests that serialize XML:
```
Recursive lookup of kotlin.collections.ArrayList
java.lang.IllegalStateException: Recursive lookup of kotlin.collections.ArrayList
	at nl.adaptivity.xmlutil.serialization.DefaultFormatCache.lookupDescriptor$serialization(DefaultFormatCache.kt:81)
	at nl.adaptivity.xmlutil.serialization.structure.XmlDescriptor$Companion.from$serialization(XmlDescriptor.kt:330)
	at nl.adaptivity.xmlutil.serialization.structure.XmlDescriptor$Companion.from$serialization$default(XmlDescriptor.kt:320)
	at nl.adaptivity.xmlutil.serialization.structure.XmlCompositeDescriptor.createElementDescriptors(XmlDescriptor.kt:1001)
...
```

It's not always in the same test and it doesn't seem to happen when running individual tests. I think this test framework runs multiple tests in parallel and they're using a single shared instance of `XML`, and so now I have to ask myself: *is* `XML` threadsafe? Because looking at [the source of the exception](https://github.com/pdvrieze/xmlutil/blob/master/serialization/src/commonMain/kotlin/nl/adaptivity/xmlutil/serialization/DefaultFormatCache.kt#L81), unless there's a threadlocal in there somewhere it seems pretty clear to me that multiple threads are going to conflict when using the same `XML`. And if it's not supposed to be threadsafe, that should be documented.

**Comment by pdvrieze** [comment, marked as answer]
The short answer is no, `XML` objects are not threadsafe (if you enable caching). I'll document it. Using different instances in different threads should be fine (but sharing the cache across invocations would improve speed).

Alternatively I've implemented the default implementation on dev to use threadlocals for Java/native (not js/wasm). However as some of the types are not public you cannot work around it except by disabling the caching.

**Reply by felixkrull-neuland** [reply]
Ok, that's at least good to know, thanks. I think we'll wait and see if the internal threadlocals work for us.

**Comment by pdvrieze** [comment]
This should be resolved/fixed with the 0.90.3 release that should be threadSafe by default. Let me know if it works/breaks. You can of course either wrap the format in a threadlocal, not use caching, or not store the format at all.
